### ISSUE TITLE
Checking for Cloudflare Captcha 

### ISSUE THREAD

**Issue by lachgil** [rootVIII/proxy_requests#15]
Thanks for providing this great script! I was about to make something similar before finding this.

I had two things I noticed when using the code; ReadTimeout was not in the list of errors 
and it could be a good idea to run a check if the returned data is not a cloudflare captcha check

I've added this bit of code to on my side and seems to do the job well;
(Added to the set_request_data function)
        if "why_captcha_headline" in req.text:
            raise CloudFlareCaptcha
then created a CloudFlareCaptcha exception class, and added it to the list of errors.

When this error is hit, it should hit the exception in each of the get/post functions, and then re-try

**Comment by rootVIII** [comment]
Do you by any chance have an example of the error in a traceback? Or if possible can you print the string/name of the error like: 
```
print(type(err).__name__ )
```

Or maybe if it's not personal you can share the URL? One of the two would help.

Thanks!

**Comment by lachgil** [comment]
Sure, details below; the string/name of the error is: ConnectionError

It only happens sometimes on some of the proxies.

Oddly, extremely long traceback was generated ( over 1000 lines) more info below


Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/urllib3/connection.py", line 157, in _new_conn
    (self._dns_host, self.port), self.timeout, **extra_kw
  File "/usr/lib/python3.7/site-packages/urllib3/util/connection.py", line 84, in create_connection
    raise err
  File "/usr/lib/python3.7/site-packages/urllib3/util/connection.py", line 74, in create_connection
    sock.connect(sa)
socket.timeout: timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/urllib3/connectionpool.py", line 662, in urlopen
    self._prepare_proxy(conn)
  File "/usr/lib/python3.7/site-packages/urllib3/connectionpool.py", line 948, in _prepare_proxy
    conn.connect()
  File "/usr/lib/python3.7/site-packages/urllib3/connection.py", line 334, in connect
    conn = self._new_conn()
  File "/usr/lib/python3.7/site-packages/urllib3/connection.py", line 164, in _new_conn
    % (self.host, self.timeout),
urllib3.exceptions.ConnectTimeoutError: (<urllib3.connection.VerifiedHTTPSConnection object at 0x7fd632ac5550>, 'Connection to 183.88.192.241 timed out. (connect timeout=4.0)')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/usr/lib/python3.7/site-packages/urllib3/connectionpool.py", line 720, in urlopen
    method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]
  File "/usr/lib/python3.7/site-packages/urllib3/util/retry.py", line 436, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='website.org.au', port=443): Max retries exceeded with url: / (Caused by ConnectTimeoutError(<urllib3.connection.VerifiedHTTPSConnection object at 0x7fd632ac5550>, 'Connection to 183.88.192.241 timed out. (connect timeout=4.0)'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/lgeeh/Documents/Surf/scraping/New/proxy_requests.py", line 67, in get
    proxies=proxies)
  File "/usr/lib/python3.7/site-packages/requests/api.py", line 75, in get
    return request('get', url, params=params, **kwargs)
  File "/usr/lib/python3.7/site-packages/requests/api.py", line 60, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/lib/python3.7/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python3.7/site-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python3.7/site-packages/requests/adapters.py", line 504, in send
    raise ConnectTimeout(e, request=request)
requests.exceptions.ConnectTimeout: HTTPSConnectionPool(host='website.org.au', port=443): Max retries exceeded with url: / (Caused by ConnectTimeoutError(<urllib3.connection.VerifiedHTTPSConnection object at 0x7fd632ac5550>, 'Connection to 183.88.192.241 timed out. (connect timeout=4.0)'))

**Comment by Mayur1995** [comment]
Thanks @rootVIII for this great module, I really appreciate this but I am running into the same issue as @lachgil when I am trying with multiple threads. For about 10 -15 sec everything works just fine but after that, I am facing the same issue as @lachgil described. 

`requests.exceptions.ConnectTimeout: HTTPSConnectionPool(host='website.org.au', port=443): Max retries exceeded with url: / (Caused by ConnectTimeoutError(<urllib3.connection.VerifiedHTTPSConnection object at 0x7fd632ac5550>, 'Connection to 183.88.192.241 timed out. (connect timeout=4.0)'))
`

Is there any workaround for this? Any help is really appreciated.

Thanks,

**Comment by rootVIII** [comment]
Sure will have a fix for that! Thanks!

**Comment by rootVIII** [comment]
@Mayur1995 is this related to a captcha? Or is this only when using multi threading?

**Comment by indiecompany** [comment]
@rootVIII errors described by @lachgil are happening to as well, I've tested versions 0.4.3 and 0.5.0, several exception inside exceptions are chained, killing the thread.

No problems with version 0.3.2 though

**Comment by hsombini** [comment]
@rootVIII errors reported by @lachgil are happening to me as well, I've tested versions 0.4.3 and 0.5.0, several exceptions are chained on some requests, killing the thread.

Version 0.3.2 works fine though

**Comment by rootVIII** [comment]
Ok gonna look at that tonight... thanks for reminding me! Btw how are you doing the init with multithreading?

**Comment by rootVIII** [comment]
well I don't know if the issue you are experiencing is related to multi threading or not?

This is kind of a stupid example but it does run:

```
from proxy_requests import ProxyRequests
from threading import Thread

def req(url):
    r = ProxyRequests(url)
    r.get()
    print(r)

if __name__ == '__main__':
    threads = []
    for i in range(1000):
        t = Thread(target=req, args=['https://www.restwords.com'])
        t.start()
        threads.append(t)
    _ = [th.join() for th in threads]


```

**Comment by rootVIII** [comment]
If there are multiple Connection Timeout errors that typically means the URL is invalid, rejecting anonymous proxies, or even your network connection could be out. Can you possibly recreate the error but with a URL that is okay to write here/share with me and the public? Thanks for the help!

**Comment by rootVIII** [comment]
I might actually just add that to the list of errors to ignore. I can't seem to recreate it but it should be fine to add it there.

**Comment by rootVIII** [comment]
I added a fix for this in 0.5.1. Can any of you let me know if you are still experiencing that same error?

**Comment by hsombini** [comment]
@rootVIII Your code is almost what I'm doing, but the request was inside a try/catch to validate the requested data, I did further analysis of the big Traceback and I found that I was trying to print the request object if an exception was caught

```
  File "main.py", line 311, in fetchModelYears
    errorLog(str(r))
  File "/usr/local/lib/python3.8/dist-packages/proxy_requests/proxy_requests.py", line 204, in __str__
    return str(self.rdata['request'])
KeyError: 'request'
```
My script was running fine on 0.3.2, but I have managed to make it run on version 0.5.0 and also on 0.5.1 removing the errorLog(str(r)), so this is not related to what @lachgil reported, my bad, jumped into conclusions due to the huge Traceback.

About the issues reported above, I'm getting data from a server that is denying most of the proxy request thanks to cloudflare stuff, I'm using a multi thread approach with several sleep/retry to not abuse proxy/server limits, there are no errors so far besides the proxy-request trying to use the same proxy that failed to get data over and over again. So I think that the above issue this may not be related to multi thread nor cloudflare protected server, since my use case covers that.
